src/gfx/drawables/mesh_tree_drawable.hpp

```diff
@@ -27,19 +23,34 @@ struct MeshTreeDrawable final : public IDrawable {
   DrawablePtr clone() const override;
   DrawableProcessorConstructor getProcessor() const override;
 
-  template <typename T> static inline void foreach (Ptr &item, T && callback) {
-    callback(item);
-    for (auto &child : item->children) {
-      MeshTreeDrawable::foreach (child, callback);
-    }
-  }
+  // Visits each MeshTreeDrawable in this tree and calls callback on it
+  template <typename T> static inline void foreach (Ptr &item, T && callback);
 
   UniqueId getId() const override { return id; }
 
   void pipelineHashCollect(detail::PipelineHashCollector &pipelineHashCollector) const override { assert(false); }
 
   bool expand(shards::pmr::vector<const IDrawable *> &outDrawables) const override;
 };
+
+template <typename T> void MeshTreeDrawable::foreach (Ptr &item, T && callback) {
+  struct Node {
+    Ptr node;
+  };
+  std::vector<Node> queue{{item}};
+
+  while (!queue.empty()) {
+    Node node = queue.back();
+    queue.pop_back();
+
+    callback(node.node);
+
+    for (auto &child : node.node->children) {
+      queue.push_back(Node{child});
+    }
```

Comment:
The order the children are processed is reversed compared to the recursive version. Does it matter?

```
// given
Root { A { a1, a2 }, B { b1, b2, C { c1, c2 } } }

// order before (recursive version)
Root -> A -> a1 -> a2 -> B -> b1 -> b2 -> C -> c1 -> c2

// order after
Root -> B -> C -> c2 -> c1 -> b2 -> b1 -> A -> a2 -> a1

```